src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -150,31 +133,164 @@ protected async ValueTask<Array> ReadArray<TAnyElement>(NpgsqlReadBuffer buf, bo
         }
 
         /// <summary>
-        /// Reads an array of element type <typeparamref name="TAnyElement"/> from the given buffer <paramref name="buf"/>.
+        /// Reads a generic list containing elements of type <typeparamref name="TRequestedElement"/> from the given buffer <paramref name="buf"/>.
         /// </summary>
-        protected async ValueTask<List<TAnyElement>> ReadList<TAnyElement>(NpgsqlReadBuffer buf, bool async)
+        protected async ValueTask<List<TRequestedElement>> ReadList<TRequestedElement>(NpgsqlReadBuffer buf, bool async)
         {
             await buf.Ensure(12, async);
             var dimensions = buf.ReadInt32();
 
             if (dimensions == 0)
-                return new List<TAnyElement>();
+                return ElementTypeInfo<TRequestedElement>.EmptyList;
             if (dimensions > 1)
-                throw new NotSupportedException($"Can't read multidimensional array as List<{typeof(TAnyElement).Name}>");
+                throw new NotSupportedException($"Can't read multidimensional array as List<{typeof(TRequestedElement).Name}>");
+
+            var containsNulls = buf.ReadInt32() == 1;
+            if (!ElementTypeInfo<TRequestedElement>.IsReferenceOrNullable && containsNulls)
+                throw UnsupportedNonNullableTypeException($"List<{typeof(TRequestedElement).Name}>", $"List<{typeof(TRequestedElement).Name}?>");
 
-            buf.ReadInt32();  // Has nulls. Not populated by PG?
             buf.ReadUInt32(); // Element OID. Ignored.
 
             await buf.Ensure(8, async);
             var length = buf.ReadInt32();
             buf.ReadInt32(); // We don't care about the lower bounds
 
-            var list = new List<TAnyElement>(length);
+            var list = new List<TRequestedElement>(length);
             for (var i = 0; i < length; i++)
-                list.Add(await _elementHandler.ReadWithLength<TAnyElement>(buf, async));
+                list.Add(await ElementHandler.ReadWithLength<TRequestedElement>(buf, async));
             return list;
         }
 
+        static InvalidOperationException UnsupportedNonNullableTypeException(string requestedType, string suggestedType)
```

Comment:
It would be better to pass types instead of strings.

Reply:
Please look at the call sites.
Do you really think it's worth to do the complex type-construction-fu that would be necessary to pass the correct type just for type safety of an internal method.
Or am I missing something obvious?

Reply:
https://github.com/npgsql/npgsql/blob/43919c0b72a7c2ad7852146df5c3f451d165fbac/src/Npgsql/TypeHandlers/ArrayHandler.cs#L85

As you can see the suggested array type has only one dimension while there could be requested a  multidimensional array. It's incorrect, so you can change exception message to omit collection type and say "Cannot read a non-nullable collection of elements because the returned array contains nulls". Otherwise, create an array first and use `GetType`.

Reply:
You're right. I think I'll go with the simplified exception message.